### PR TITLE
Move add_date/mod_date from Biosample and DataGeneration into provenance_metadata (range: datetime)

### LINKED ISSUE #2829: Add add_date and mod_date slots to ProvenanceMetadata class
## Context

Part of #2806. Per Alicia's Feb 4 comment:

> mod_date and add_date will get moved to ProvenanceMetadata, provenance_metadata slot needs to apply to more classes. Update mod_date and add_date to an appropriate range (ISO date format). This will need a migrator.

## Scope

Add `add_date` and `mod_date` to the `ProvenanceMetadata` class in `src/schema/basic_classes.yaml`.

Currently `ProvenanceMetadata` has: `git_url`, `version`, `source_system_of_record`, `submission_portal_identifier`, `type`.

This issue is schema-only — range/pattern constraints are #2830 and the migrator is #2831.

## Current state of add_date / mod_date

- Defined in `src/schema/basic_slots.yaml` (lines 346-352)
- `range: string`, no pattern
- Asserted directly on Biosample and DataGeneration
- Inherited by MassSpectrometry, NucleotideSequencing

## Acceptance criteria

- [ ] `add_date` and `mod_date` listed in ProvenanceMetadata slots
- [ ] Schema validates (`linkml-lint`)

### LINKED ISSUE #2830: Set date range constraint on add_date and mod_date
## Context

Part of #2806. Currently `add_date` and `mod_date` are `range: string` with no pattern.

## Terminology

These terms describe the same date format at different layers:

| Term | Layer | Notes |
|------|-------|-------|
| `linkml:date` | [LinkML built-in type](https://linkml.io/linkml-model/latest/docs/date/) | What you write in schema YAML: `range: date` |
| `xsd:date` | [XML Schema Datatype](https://www.w3.org/TR/xmlschema-2/#date) | The URI that `linkml:date` maps to |
| [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339#section-5.6) `full-date` | Internet standard | What JSON Schema `"format": "date"` references |
| [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) | International standard | The broad standard that defines `YYYY-MM-DD` |
| `YYYY-MM-DD` | The pattern | e.g., `2021-03-31` |

**Chain:** `range: date` (linkml:date) → xsd:date → JSON Schema `"format": "date"` (RFC 3339) → validates `YYYY-MM-DD` strings.

## Data in MongoDB (localhost, nmdc database)

- **add_date**: always `YYYY-MM-DD` (2,688 biosamples, 5,285 data generations)
- **mod_date**: mixed — mostly `YYYY-MM-DDT00:00:00` but some plain `YYYY-MM-DD` (biosample: 2,688, data generation: 5,236)

The `T00:00:00` on mod_date is likely a GOLD ingest artifact.

## Decision: `range: date` (linkml:date, YYYY-MM-DD)

Implemented via `slot_usage` on ProvenanceMetadata (PR #2834). The global range remains `string` for backward compatibility.

### Time-of-day analysis (2026-02-18)

Queried all distinct `add_date` and `mod_date` values in MongoDB to determine whether any meaningful time-of-day information exists:

| Collection | Slot | Total distinct values | With `T` time component | Non-midnight time |
|------------|------|-----------------------|------------------------|-------------------|
| biosample_set | add_date | 43 | 19 (all `T00:00:00`) | **0** |
| biosample_set | mod_date | 33 | 18 (all `T00:00:00`) | **0** |
| data_generation_set | add_date | 67 | 21 (all `T00:00:00`) | **0** |
| data_generation_set | mod_date | 51 | 22 | **1** (`2024-11-07T15:02:18`) |

The single non-midnight value (`2024-11-07T15:02:18`) appears on **2,633 data_generation documents** — a bulk batch timestamp from an ETL run, not per-record precision. Knowing it was Nov 7, 2024 is sufficient.

**Conclusion:** Migrating to date-only (`YYYY-MM-DD`) loses zero meaningful time-of-day information. `range: date` (linkml:date) is appropriate for both slots.

## Options considered

1. ✅ `range: date` (linkml:date) — cleanest. Requires normalizing mod_date values that have `T00:00:00` (in migrator, see #2831).
2. `range: string` with pattern `^\d{4}-\d{2}-\d{2}$` — simpler but less semantically precise.
3. `range: date_or_datetime` — accommodates the mixed formats but doesn't clean them up.

These are system-generated dates (from GOLD, submission portal, nmdc-runtime). Partial dates are not needed, so formal types are appropriate.

Note: `started_at_time` and `ended_at_time` already have commented-out `# range: datetime` — suggesting formal types were the original intent.

## YAML quoting requirement

Date values in YAML data files must be quoted (e.g., `'2021-03-31'`). Without quotes, PyYAML (YAML 1.1) parses bare dates as native `datetime.date` objects, which `linkml-validate`'s JSON Schema pathway rejects. LinkML's `ReferenceValidator` and `XSDDate` class do support native date objects, but that code path is not used by `linkml-validate`.

## Acceptance criteria

- [x] `add_date` and `mod_date` constrained to `range: date` (linkml:date) via `slot_usage` on ProvenanceMetadata
- [x] Schema validates
- [x] Valid and invalid example data files added

### LINKED ISSUE #2831: Migrator: move Biosample add_date/mod_date into provenance_metadata
## Context

Part of #2806. After #2829 (add slots to ProvenanceMetadata) and #2830 (set range constraint), write a migrator that moves `add_date` and `mod_date` values from top-level Biosample fields into the embedded `provenance_metadata` object.

## Scope

**Biosample only** — DataGeneration migration is a separate issue (#2832).

### Schema changes (in the same PR as the migrator)

- Remove `add_date` and `mod_date` from the Biosample class `slots` list in `src/schema/core.yaml`
- These slots remain defined globally in `basic_slots.yaml` (still used by DataGeneration)
- Update any valid/invalid test data files that use top-level `add_date`/`mod_date` on Biosample

### Data migration

For each document in `biosample_set`:
1. Read `add_date` and `mod_date` from the top-level document
2. If `provenance_metadata` doesn't exist, create it (with `type: "nmdc:ProvenanceMetadata"`)
3. Set `provenance_metadata.add_date` and `provenance_metadata.mod_date`
4. Normalize mod_date values: strip `T00:00:00` suffix to plain `YYYY-MM-DD` (linkml:date format). Note: biosample_set has zero non-midnight time values, so no information is lost (see #2830).
5. Remove top-level `add_date` and `mod_date`

## Data volumes

- 2,688 biosamples have `add_date` and `mod_date`
- add_date is always `YYYY-MM-DD`
- mod_date is mostly `YYYY-MM-DDT00:00:00`, some plain `YYYY-MM-DD`
- Zero non-midnight time values in biosample_set (see time-of-day analysis in #2830)

## Cross-repo impact

Moving biosample `add_date`/`mod_date` into `provenance_metadata` affects code in other repos. These changes should be coordinated with the schema migration.

### nmdc-runtime — GOLD translator (writes these fields)

| File | What it does | Change needed |
|------|-------------|---------------|
| `nmdc_runtime/site/translation/gold_translator.py` `_translate_biosample` (lines 568-601) | Sets `add_date=gold_biosample["addDate"]` and `mod_date` (with fallback logic) directly on Biosample | Write into `provenance_metadata.add_date` / `provenance_metadata.mod_date` instead |
| `nmdc_runtime/site/translation/gold_translator.py` `_get_mod_date` (lines 147-160) | Fallback: if `modDate` is null, returns `addDate` | No change needed (generic helper), but callers need updating |

Note: `submission_portal_translator.py` does **not** set `add_date`/`mod_date` on biosamples — only on NucleotideSequencing objects.

### nmdc-server — Data portal (reads these fields)

| File | What it does | Change needed |
|------|-------------|---------------|
| `nmdc_server/models.py` Biosample class (lines 380-381) | PostgreSQL `add_date`/`mod_date` DateTime columns | Read from `provenance_metadata` during ingest, or keep as denormalized columns |
| `nmdc_server/ingest/biosample.py` `coerce_date` validator | Parses GOLD-format dates (`"30-OCT-14 12.00.00..."`) to ISO during MongoDB→PostgreSQL ETL | Update to read from `provenance_metadata.add_date`/`mod_date`; GOLD-format coercion may no longer be needed if migrator normalizes values |
| `nmdc_server/schemas.py` BiosampleBase | API serialization exposes `add_date`/`mod_date` as `Optional[DateType]` | Update source to `provenance_metadata` or keep as API fields populated from new location |
| `web/src/data/api.ts` BiosampleSearchResult | TypeScript interface includes `add_date: string` and `mod_date: string` | May need update depending on API changes |
| `web/src/encoding.ts` | Field metadata — both fields are `hideFacet: true, hideAttr: true` | Currently hidden from UI; minimal user-facing impact |
| `tests/test_query.py` `test_between_query_column` (lines 209-225) | Tests `between` range queries on biosample `add_date` | Update test data and possibly query paths |
| `nmdc_server/fakes.py` BiosampleFactory | Generates fake `add_date`/`mod_date` for tests | Update to match new structure |
| `migrations/780ab86a4f27` (squashed initial) | Creates `add_date`/`mod_date` columns in biosample table | New migration needed if columns are removed/renamed |

### Other repos (minor)

| Repo | File | What it does | Change needed |
|------|------|-------------|---------------|
| ms_material_processing_curation | `materialprocessing_metadata.py` (line 283) | Reads biosample `add_date` into a dataframe for curation | Update field path |
| external-metadata-awareness | `normalize_biosample_measurements.py` (line 103) | Lists `add_date`/`mod_date` as biosample field names (likely exclusion list) | Update field names |
| sample-annotator (deprecated) | `gold_nmdc_pipeline.py` | Predecessor to nmdc-runtime translator | No action (archived) |
| pilot (deprecated) | `encoding.js` | Predecessor to nmdc-server encoding | No action (archived) |

### Key observations

- **GOLD translation is the primary source** of biosample `add_date`/`mod_date` values
- **Submission portal does not set these on biosamples** — only on NucleotideSequencing
- **Data portal hides these fields** from faceted search and attribute displays, but supports `between` range queries on `add_date`
- **No MongoDB queries** filter on `biosample_set.add_date`/`mod_date` directly — they are only read during ETL

## Migrator pattern

Extend `MigratorBase`, use `self.adapter.process_each_document("biosample_set", [callback])` where callback is `dict → dict`. See `migrator_from_10_3_0_to_10_4_0.py` for a recent example.

## Acceptance criteria

- [ ] `add_date` and `mod_date` removed from Biosample class slots
- [ ] Migrator moves add_date/mod_date into provenance_metadata for all biosamples
- [ ] mod_date values normalized to YYYY-MM-DD
- [ ] Top-level add_date/mod_date removed from biosample documents
- [ ] Valid/invalid test data files updated
- [ ] Doctests pass
- [ ] Cross-repo changes identified and tracked (see table above)

### LINKED ISSUE #2860: Move add_date/mod_date from DataGeneration into provenance_metadata
## Context

PR #2833 moves `add_date` and `mod_date` from Biosample into `provenance_metadata` with `range: date`. This issue tracks doing the same for **DataGeneration**.

Part of the provenance metadata consolidation (#2806, #2829).

## Problem

DataGeneration currently has `add_date` and `mod_date` as direct slots (basic_classes.yaml:420,424) but does NOT have a `provenance_metadata` slot. To consolidate provenance metadata, these should move into `provenance_metadata` like we're doing for Biosample.

## Complication: GOLD-format dates

DataGeneration records in production MongoDB contain GOLD-format timestamps like `30-OCT-14 12.00.00.000000000 AM`. These are NOT valid under `range: date` (xsd:date, YYYY-MM-DD) or `range: date_or_datetime`.

The migrator must normalize these to ISO format (e.g., `2014-10-30` or `2014-10-30T00:00:00`) when moving them into `provenance_metadata`. The test data files also need updating:
- 20 GOLD-format date values across DataGeneration test data files
- 1 ISO date, 2 ISO datetimes (already valid)

## Schema changes needed

1. Add `provenance_metadata` slot to DataGeneration class
2. Remove `add_date` and `mod_date` from DataGeneration's slots list
3. (slot_usage with `range: date` on ProvenanceMetadata is already done by #2833)

## Migrator changes

Extend the migrator to process `data_generation_set`:
- Move `add_date`/`mod_date` into `provenance_metadata`
- Normalize GOLD-format dates to ISO 8601 date strings

## Test data files to update

Files with DataGeneration `add_date`/`mod_date`:
- Database-data-generations.yaml
- Database-gen_from_omics.yaml
- Database-mass-spectrometry.yaml
- Database-metabolomics_calibration_example.yaml
- Database-nom_calibration_example.yaml
- Database-multiple-paths.yaml
- Database-GOLD_amplicon.yaml
- Database-nmdc-example.yaml
- NucleotideSequencing-*.yaml

## Related

- #2806 — Provenance squad epic
- #2829 — Move dates to provenance_metadata (Biosample, done in #2833)
- #2832 — Originally tracked DataGeneration scope

### PR DESCRIPTION
**Pull request by turbomam**
## Summary

Consolidates `add_date` and `mod_date` into the `ProvenanceMetadata` class for both **Biosample** and **DataGeneration**, with a `range: datetime` constraint (xsd:dateTime, RFC 3339). Includes a structural migrator with GOLD-format date normalization and updated test data.

- **Schema**: Add `add_date`/`mod_date` to ProvenanceMetadata slots with `slot_usage` constraining `range: datetime`; remove both from Biosample (core.yaml) and DataGeneration (basic_classes.yaml); add `provenance_metadata` slot to DataGeneration; add class-specific `provenance_metadata` slot_usage descriptions on Study, DataGeneration, and Biosample
- **Migrator**: `migrator_from_11_16_1_to_11_17_0.py` moves top-level dates into `provenance_metadata` on both `biosample_set` and `data_generation_set` documents, normalizing GOLD-format timestamps to RFC 3339 datetime strings (22 doctests)
- **Test data**: Updated 15 existing files, added 1 valid + 2 invalid test files, regenerated `Database-interleaved.yaml`

## Why `range: datetime`

We evaluated multiple options:

| Type | JSON Schema output | Validates correctly? |
|------|-------------------|---------------------|
| `datetime` (linkml:datetime → xsd:dateTime) | `{"format": "date-time"}` | Yes — requires RFC 3339 with timezone designator |
| `date` (linkml:date → xsd:date) | `{"format": "date"}` | Yes but too coarse-grained for comparing events |
| `date_or_datetime` (linkml:DateOrDatetime) | `{"type": "string"}` (no format) | **No** — accepts anything, GOLD-format passes silently |

Consensus from Patrick, Eric, and the metadata meeting: `datetime` provides unambiguous point-in-time comparisons and is natively supported by JSON Schema. Records are populated by `nmdc-runtime` automation, so generating full datetimes adds no burden.

## GOLD date normalization

The migrator includes a `normalize_date_to_datetime()` function that handles multiple input formats:
- **GOLD timestamps** (`30-OCT-14 12.00.00.000000000 AM`) → `2014-10-30T00:00:00Z`
- **ISO datetimes** (`2024-11-07T15:02:18`) → `2024-11-07T15:02:18Z`
- **ISO datetimes with offset** (`2024-11-07T15:02:18+00:00`) → preserved as-is
- **ISO dates** (`2021-03-31`) → `2021-03-31T00:00:00Z`

## Legacy date semantics

Biosample and NucleotideSequencing records brought in from GOLD before 2026-03-23 may carry GOLD-sourced dates rather than NMDC-asserted timestamps. Going forward, `add_date`/`mod_date` reflect when records were added to or modified in the NMDC database. Patrick will update the GOLD translator in nmdc-runtime to assert NMDC timestamps.

### Schema
| File | Change |
|------|--------|
| `src/schema/basic_classes.yaml` | Add slots + slot_usage on ProvenanceMetadata; add `provenance_metadata` slot_usage on Study and DataGeneration; remove `add_date`/`mod_date` from DataGeneration |
| `src/schema/basic_slots.yaml` | Simplify global `add_date`/`mod_date` descriptions to generic, database-agnostic language |
| `src/schema/core.yaml` | Remove `add_date`/`mod_date` from Biosample; add `provenance_metadata` slot_usage on Biosample |

### Migrator
| File | Change |
|------|--------|
| `nmdc_schema/migrators/migrator_from_11_16_1_to_11_17_0.py` | **New** — structural migrator with GOLD date normalizer (22 doctests) |

### CLAUDE.md
| File | Change |
|------|--------|
| `CLAUDE.md` | Clarify derived files guidance; add example data guidelines |

## Validation

- Migrator doctests pass (22 tests covering both collections, all date formats, edge cases including timezone offsets)
- `make squeaky-clean all test` passes on branch
- **Tested against production data**: Exported 15,440 biosamples + 10,891 data generations from `api.microbiomedata.org`, ran migrator, validated result — **"No issues found"**

## Required follow-up

- #2861 — Update nmdc-runtime GOLD/NEON ingests to write dates into provenance_metadata (@sujaypatil96)
- #2862 — Constrain global range of add_date/mod_date from string to datetime (blocked by #2861)
- #2863 — Create a skill/make target for validating invalid example files

## Closes

Closes #2829, Closes #2830, Closes #2831, Closes #2860
Part of #2806

Supersedes #2834 (its changes are folded into this PR)

### CODE REVIEW

**Comment by pkalita-lbl** [comment]
I would argue that `range: datetime` is a better option for these slots. 

* Why not `range: date`? I think `date` is too course-grained. I think there are going to be times when we want to know whether a record was modified before or after some event. Just having the date recorded may not be enough resolution to answer that question.
* Why not `range: date_or_datetime`? As the name implies, this would permit a date or a datetime. Values where a date is used still leave us with the same course-grained issue noted above. Plus, since `date_or_datetime` is sort of a thing that LinkML made up, it's not supported by JSON Schema natively. So we'd have to add a custom regex pattern ourselves or bake one into the LinkML JSON Schema generator. 

On the other hand `range: datetime` solves both of these problems. It identifies unique points in time that can be compared unambiguously with other points in time. It also is natively supported by JSON Schema so we don't have to do any extra work to have proper validation. These slots are, generally speaking, going to be populated in an automated way by `nmdc-runtime`, so there's no extra work to populate a full datetime. In fact, it might be slightly _more_ work to truncate a datetime into a date before populating these slots.

Yes, there are some existing values in MongoDB that are pure dates. We'd have to migrate those to some artificial midnight UTC time (presumably). If our options are:

1. Model this as what we really want (datetimes) and all new data coming in adheres to it at the cost of having to "invent" a few times
2. Model this in the lowest-common-denomonator way (dates) and have an "unopinionated" migrator at the cost of forever having lower resolution data

I'd pick 1 every time.

**Comment by eecavanna** [comment]
I'm also in favor of option 1 (`datetime`).

I'm reviewing this migrator now as-is (knowing that the PR may eventually be updated to use `datetime` instaed of `date`, at which point, I'd want to review it one more time). I will "think out loud" (via comments) as I'm reviewing this migrator because I know that this is the migrator author's first time writing a migrator. 👍

File: nmdc_schema/migrators/migrator_from_11_16_1_to_11_17_0.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    # GOLD-format: DD-MON-YY HH.MM.SS.NNNNNNNNN AM/PM
+    # The fractional seconds field is 9 digits; strptime %f handles 6 digits,
+    # so we treat the last 3 as literal zeros in the format string.
+    try:
+        dt = datetime.strptime(value, _GOLD_DATE_FORMAT)
+        return dt.strftime("%Y-%m-%d")
+    except ValueError:
+        # Return as-is if we can't parse — the validator will catch it
+        return value
```
Comment (eecavanna) [inline nmdc_schema/migrators/migrator_from_11_16_1_to_11_17_0.py]: I would prefer that the migrator raise an exception in this case (fail early).

**Review by eecavanna — COMMENTED** [review]
The migrator is coming along! I left some comments that stem from the fact that the "initial" schema (i.e. 11.16.1) allows `add_date` and `mod_date` to contain of any arbitrary string, while I don't think the migrator accounts for all arbitrary strings. I can walk you through my feedback later this week (or we can wait until the question of `date` versus `datetime` has been settled).

**Comment by pkalita-lbl** [comment]
Can we also clarify in the slot descriptions that `add_date` means that it's the date the record was added to the **NMDC** database and `mod_date` is the date the record was updated in the **NMDC** database.

**Comment by turbomam** [comment]
+1 to @pkalita-lbl's point about clarifying the slot descriptions. These dates should represent when the record was added to or modified within NMDC, not when it was created in the source system (e.g. GOLD).

For context, the GOLD translator currently passes through Gold's `addDate` and `modDate` values as-is (see `gold_translator.py` lines 570, 601, 663-664). The submission portal translator already uses "today" — i.e. the NMDC-centric interpretation. Adopting @pkalita-lbl's semantics means the GOLD ingest would need to assert NMDC timestamps instead of passing through Gold's dates.

Gold's original `addDate`/`modDate` values will remain accessible via the JGI data lakehouse (GOLD biosample tables in Dremio), so we won't lose that provenance — it just won't live in the NMDC schema.

Related issues:
- #8 — "Do we keep GOLD add_date and mod_date?" (closed 2024, concluded yes for sync tracking — but that rationale changes if we redefine these as NMDC-asserted dates)
- #884 — "Consistency among all datetime fields" (open since 2023, broader scope)

**Comment by turbomam** [comment]
Apologies — the re-review request went out before the code was pushed. The changes described in my previous comment are now live on the branch (7d6f26ca8).

File: nmdc_schema/migrators/migrator_from_11_16_1_to_11_17_0.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+_GOLD_DATE_FORMAT = "%d-%b-%y %I.%M.%S.%f000 %p"
+
+_ISO_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
+_ISO_DATETIME_RE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}")
+
+
+def normalize_date_to_datetime(value: str) -> str:
+    r"""
+    Normalizes a date string to RFC 3339 datetime format (YYYY-MM-DDTHH:MM:SSZ).
+
```
Comment (eecavanna) [inline nmdc_schema/migrators/migrator_from_11_16_1_to_11_17_0.py]:
```suggestion
    Normalizes a date string to RFC 3339 datetime format (YYYY-MM-DDTHH:MM:SSZ),
    setting—not converting—its time zone to UTC if it lacks time zone info.
```

File: src/schema/basic_slots.yaml [hunk 3]
```diff
@@ -345,11 +345,17 @@ slots:
 
   add_date:
     range: string
-    description: The date on which the information was added to the database.
+    description: >-
+      The date and time at which a record was added to the NMDC database.
+      Note: legacy records may contain Gold-sourced dates that predate this
+      convention; going forward this slot reflects NMDC-internal provenance.
```
Comment (pkalita-lbl) [inline src/schema/basic_slots.yaml]: Thoughts about about the "legacy records" note:

- It would be better as a `comment` in the `ProvenanceMetadata` `slot_usage` for `add_date` and `mod_date` since there already is a different note about "legacy records" there.
- It might be good to note that "legacy records" in this case means `Biosample` and `NucleotideSequencing` records that were brought in from GOLD before 2026-03-23 (the date we'll presumably run this migration on production data).
- The existing note in the `ProvenanceMetadata` `slot_usage` talks about "legacy records" migrated from date-only values. In that case "legacy records" means records from any source created or updated before 2026-03-23.

File: src/schema/basic_classes.yaml [hunk 4]
```diff
@@ -522,12 +521,24 @@ classes:
...
     slot_usage:
+      add_date:
+        range: datetime
+        comments:
+          - "range: datetime is the LinkML datetime type (linkml:datetime, mapped to xsd:dateTime). JSON Schema validates this as RFC 3339 date-time, which requires a timezone designator (e.g. 2021-03-31T00:00:00Z). Constrained here within ProvenanceMetadata; the global range remains string for backward compatibility with legacy exported data."
+          - Legacy records migrated from date-only values use a conventional T00:00:00Z time component, which should not be interpreted as an actual time of day.
+      mod_date:
+        range: datetime
+        comments:
+          - "range: datetime is the LinkML datetime type (linkml:datetime, mapped to xsd:dateTime). JSON Schema validates this as RFC 3339 date-time, which requires a timezone designator (e.g. 2021-03-31T00:00:00Z). Constrained here within ProvenanceMetadata; the global range remains string for backward compatibility with legacy exported data."
```
Comment (pkalita-lbl) [inline src/schema/basic_classes.yaml]: This feels a bit more of an internal-facing comment? If that's the case I wouldn't include it in `comments`, which is intended to be public-facing (i.e. shown in the doc pages).
